{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Installation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`conda install bioconductor-genomicranges bioconductor-summarizedexperiment -y`  \n",
    "\n",
    "`R`  \n",
    "`devtools::install_github(\"caleblareau/BuenColors\")`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colMeans,\n",
      "    colnames, colSums, dirname, do.call, duplicated, eval, evalq,\n",
      "    Filter, Find, get, grep, grepl, intersect, is.unsorted, lapply,\n",
      "    lengths, Map, mapply, match, mget, order, paste, pmax, pmax.int,\n",
      "    pmin, pmin.int, Position, rank, rbind, Reduce, rowMeans, rownames,\n",
      "    rowSums, sapply, setdiff, sort, table, tapply, union, unique,\n",
      "    unsplit, which, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Loading required package: IRanges\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    shift\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "Loading required package: MASS\n",
      "\n",
      "Attaching package: ‘MASS’\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Attaching package: ‘Matrix’\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(GenomicRanges)\n",
    "library(SummarizedExperiment)\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "library(BuenColors)\n",
    "library(Matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preprocess"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`bsub < count_reads_peaks_erisone.sh`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5335"
      ],
      "text/latex": [
       "5335"
      ],
      "text/markdown": [
       "5335"
      ],
      "text/plain": [
       "[1] 5335"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "path = './count_reads_peaks_output/'\n",
    "files <- list.files(path,pattern = \"\\\\.txt$\")\n",
    "length(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#assuming tab separated values with a header    \n",
    "datalist = lapply(files, function(x)fread(paste0(path,x))$V4) \n",
    "#assuming the same header/columns for all files\n",
    "datafr = do.call(\"cbind\", datalist) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>97998</li>\n",
       "\t<li>5335</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 97998\n",
       "\\item 5335\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 97998\n",
       "2. 5335\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 97998  5335"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(datafr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_regions = read.csv(\"../../input/atac_v1_pbmc_5k_peaks.bed\",\n",
    "                      sep = '\\t',header=FALSE,stringsAsFactors=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>97998</li>\n",
       "\t<li>3</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 97998\n",
       "\\item 3\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 97998\n",
       "2. 3\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 97998     3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(df_regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaknames = paste(df_regions$V1,df_regions$V2,df_regions$V3,sep = \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'chr1_10244_10510'</li>\n",
       "\t<li>'chr1_237575_237942'</li>\n",
       "\t<li>'chr1_565098_565554'</li>\n",
       "\t<li>'chr1_569172_569645'</li>\n",
       "\t<li>'chr1_713421_715095'</li>\n",
       "\t<li>'chr1_752386_753061'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'chr1\\_10244\\_10510'\n",
       "\\item 'chr1\\_237575\\_237942'\n",
       "\\item 'chr1\\_565098\\_565554'\n",
       "\\item 'chr1\\_569172\\_569645'\n",
       "\\item 'chr1\\_713421\\_715095'\n",
       "\\item 'chr1\\_752386\\_753061'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'chr1_10244_10510'\n",
       "2. 'chr1_237575_237942'\n",
       "3. 'chr1_565098_565554'\n",
       "4. 'chr1_569172_569645'\n",
       "5. 'chr1_713421_715095'\n",
       "6. 'chr1_752386_753061'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"chr1_10244_10510\"   \"chr1_237575_237942\" \"chr1_565098_565554\"\n",
       "[4] \"chr1_569172_569645\" \"chr1_713421_715095\" \"chr1_752386_753061\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(peaknames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'AAACGAAAGCGCAATG-1'</li>\n",
       "\t<li>'AAACGAAAGGGTATCG-1'</li>\n",
       "\t<li>'AAACGAAAGTAACATG-1'</li>\n",
       "\t<li>'AAACGAAAGTTACACC-1'</li>\n",
       "\t<li>'AAACGAACAGAGATGC-1'</li>\n",
       "\t<li>'AAACGAACATGCTATG-1'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'AAACGAAAGCGCAATG-1'\n",
       "\\item 'AAACGAAAGGGTATCG-1'\n",
       "\\item 'AAACGAAAGTAACATG-1'\n",
       "\\item 'AAACGAAAGTTACACC-1'\n",
       "\\item 'AAACGAACAGAGATGC-1'\n",
       "\\item 'AAACGAACATGCTATG-1'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'AAACGAAAGCGCAATG-1'\n",
       "2. 'AAACGAAAGGGTATCG-1'\n",
       "3. 'AAACGAAAGTAACATG-1'\n",
       "4. 'AAACGAAAGTTACACC-1'\n",
       "5. 'AAACGAACAGAGATGC-1'\n",
       "6. 'AAACGAACATGCTATG-1'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"AAACGAAAGCGCAATG-1\" \"AAACGAAAGGGTATCG-1\" \"AAACGAAAGTAACATG-1\"\n",
       "[4] \"AAACGAAAGTTACACC-1\" \"AAACGAACAGAGATGC-1\" \"AAACGAACATGCTATG-1\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(sapply(strsplit(files,'\\\\.'),'[', 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(datafr) = sapply(strsplit(files,'\\\\.'),'[', 2)\n",
    "rownames(datafr) = peaknames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 3 × 3 of type int</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>AAACGAAAGCGCAATG-1</th><th scope=col>AAACGAAAGGGTATCG-1</th><th scope=col>AAACGAAAGTAACATG-1</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>chr1_10244_10510</th><td>1</td><td>0</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>chr1_237575_237942</th><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>chr1_565098_565554</th><td>0</td><td>0</td><td>6</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 3 × 3 of type int\n",
       "\\begin{tabular}{r|lll}\n",
       "  & AAACGAAAGCGCAATG-1 & AAACGAAAGGGTATCG-1 & AAACGAAAGTAACATG-1\\\\\n",
       "\\hline\n",
       "\tchr1\\_10244\\_10510 & 1 & 0 & 1\\\\\n",
       "\tchr1\\_237575\\_237942 & 0 & 0 & 0\\\\\n",
       "\tchr1\\_565098\\_565554 & 0 & 0 & 6\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 3 × 3 of type int\n",
       "\n",
       "| <!--/--> | AAACGAAAGCGCAATG-1 | AAACGAAAGGGTATCG-1 | AAACGAAAGTAACATG-1 |\n",
       "|---|---|---|---|\n",
       "| chr1_10244_10510 | 1 | 0 | 1 |\n",
       "| chr1_237575_237942 | 0 | 0 | 0 |\n",
       "| chr1_565098_565554 | 0 | 0 | 6 |\n",
       "\n"
      ],
      "text/plain": [
       "                   AAACGAAAGCGCAATG-1 AAACGAAAGGGTATCG-1 AAACGAAAGTAACATG-1\n",
       "chr1_10244_10510   1                  0                  1                 \n",
       "chr1_237575_237942 0                  0                  0                 \n",
       "chr1_565098_565554 0                  0                  6                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "datafr[1:3,1:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>97998</li>\n",
       "\t<li>5335</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 97998\n",
       "\\item 5335\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 97998\n",
       "2. 5335\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 97998  5335"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(datafr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# saveRDS(datafr, file = './datafr.rds')\n",
    "# datafr = readRDS('./datafr.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Obtain Feature Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time <- Sys.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "set.seed(2019)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata <- read.table('../../input/metadata.tsv',\n",
    "                         header = TRUE,\n",
    "                         stringsAsFactors=FALSE,quote=\"\",row.names=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import counts\n",
    "counts <- data.matrix(datafr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>97998</li>\n",
       "\t<li>5335</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 97998\n",
       "\\item 5335\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 97998\n",
       "2. 5335\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 97998  5335"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 3 × 3 of type int</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>AAACGAAAGCGCAATG-1</th><th scope=col>AAACGAAAGGGTATCG-1</th><th scope=col>AAACGAAAGTAACATG-1</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>chr1_10244_10510</th><td>1</td><td>0</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>chr1_237575_237942</th><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>chr1_565098_565554</th><td>0</td><td>0</td><td>6</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 3 × 3 of type int\n",
       "\\begin{tabular}{r|lll}\n",
       "  & AAACGAAAGCGCAATG-1 & AAACGAAAGGGTATCG-1 & AAACGAAAGTAACATG-1\\\\\n",
       "\\hline\n",
       "\tchr1\\_10244\\_10510 & 1 & 0 & 1\\\\\n",
       "\tchr1\\_237575\\_237942 & 0 & 0 & 0\\\\\n",
       "\tchr1\\_565098\\_565554 & 0 & 0 & 6\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 3 × 3 of type int\n",
       "\n",
       "| <!--/--> | AAACGAAAGCGCAATG-1 | AAACGAAAGGGTATCG-1 | AAACGAAAGTAACATG-1 |\n",
       "|---|---|---|---|\n",
       "| chr1_10244_10510 | 1 | 0 | 1 |\n",
       "| chr1_237575_237942 | 0 | 0 | 0 |\n",
       "| chr1_565098_565554 | 0 | 0 | 6 |\n",
       "\n"
      ],
      "text/plain": [
       "                   AAACGAAAGCGCAATG-1 AAACGAAAGGGTATCG-1 AAACGAAAGTAACATG-1\n",
       "chr1_10244_10510   1                  0                  1                 \n",
       "chr1_237575_237942 0                  0                  0                 \n",
       "chr1_565098_565554 0                  0                  6                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(counts)\n",
    "counts[1:3,1:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>23938</li>\n",
       "\t<li>5</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 23938\n",
       "\\item 5\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 23938\n",
       "2. 5\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 23938     5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 3 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>V2</th><th scope=col>V3</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr19</td><td>58858171</td><td>58864865</td></tr>\n",
       "\t<tr><td>chr19</td><td>58863335</td><td>58866549</td></tr>\n",
       "\t<tr><td>chr10</td><td>52559168</td><td>52645435</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       " V1 & V2 & V3\\\\\n",
       " <chr> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t chr19 & 58858171 & 58864865\\\\\n",
       "\t chr19 & 58863335 & 58866549\\\\\n",
       "\t chr10 & 52559168 & 52645435\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 3\n",
       "\n",
       "| V1 &lt;chr&gt; | V2 &lt;int&gt; | V3 &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| chr19 | 58858171 | 58864865 |\n",
       "| chr19 | 58863335 | 58866549 |\n",
       "| chr10 | 52559168 | 52645435 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1    V2       V3      \n",
       "1 chr19 58858171 58864865\n",
       "2 chr19 58863335 58866549\n",
       "3 chr10 52559168 52645435"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# import gene bodies; restrict to TSS\n",
    "gdf <- read.table(\"../../input/hg19/hg19-tss.bed\", stringsAsFactors = FALSE)\n",
    "dim(gdf)\n",
    "gdf[1:3,1:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss <- data.frame(chr = gdf$V1, gene = gdf$V4, stringsAsFactors = FALSE)\n",
    "tss$tss <-  ifelse(gdf$V5 == \"+\", gdf$V3, gdf$V2)\n",
    "tss$start <- ifelse(tss$tss - 50000 > 0, tss$tss - 50000, 0)\n",
    "tss$stop <- tss$tss + 50000\n",
    "tss_idx <- makeGRangesFromDataFrame(tss, keep.extra.columns = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import ATAC peaks\n",
    "\n",
    "adf <- data.frame(fread('../../input/atac_v1_pbmc_5k_peaks.bed'))\n",
    "colnames(adf) <- c(\"chr\", \"start\", \"end\")\n",
    "adf$mp <- (adf$start + adf$end)/2\n",
    "atacgranges <- makeGRangesFromDataFrame(adf, start.field = \"mp\", end.field = \"mp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find overlap between ATAC peaks and Ranges linker\n",
    "ov <- findOverlaps(atacgranges, tss_idx) #(query, subject)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Don't know how to automatically pick scale for object of type table. Defaulting to continuous.\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWgAAAFoCAMAAABNO5HnAAAC9FBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUW\nFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJyco\nKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7\nOzs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExN\nTU1OTk5PT09QUFBSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19g\nYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFy\ncnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OE\nhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWW\nlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eo\nqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi6urq7\nu7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnLy8vMzMzNzc3O\nzs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g\n4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy\n8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///8Xt2AlAAAACXBIWXMA\nABJ0AAASdAHeZh94AAARyUlEQVR4nO2de3wU1dnHnyUkm4SQcAkNEik3LxAFXiMi+kKUiyDB\nRIOvGgQVEAG5JnnVIkgpEiq1YsjbVysiWMBr85KWQKSUi6AiVKBgUkCMvCkXTeRSFISQnH86\nZ3Z2dyaZPdlhzj7Z3Tzfz8eZ4cw5v33mm83sTDLxACNQgKYuoLlAopEg0UiQaCRINBIkGgkS\njYRt0cthvmvjDNzEWBac1O/ctdVuvJH/6+XsdMjPvvVKaWoCKrq0xXK78QaqoiMGD7/gZ+fw\nFn3mZK1u32qQK/pjeMr/zuEt2ohs0X+Bhf53Dm/R/OhqX+wT22bwRsbGgEI5Y98+3TkyedJx\n3ulodlKr4Qdv6s9Yepft3WP+i7F9jyRHxg8sVvZlJpffHx93z6HqiR3ih5V5X8EzvC/Pe15r\nzWz/xeDYxIeO8O2zed2ikqd+xze9carobdHJX3kLEg51F+TCW6iuS2bSVw+2ixlQWu81/SIA\nomdB2jOT2zo2sE3jILvwDDuWDINmDIZrjirlJzlGTe3eJpGLbtuu/7B5bFdM67F5D7R0bFIO\nI67tw6umwfU3DVs+L/K6y+4X8A5/bxqMKtylNWc6Wz+2493OiYeVV06Bgf/9YES3U0wfx0vZ\nE5/0D+YtSDhUK8iFrlBdl8xWP7t3+ZL2EXsNrWiiu6Sp3KmJvtDybqX5gGOE+9QxCgqV5Wsw\nmLH7YQ1j5wcCFw3j+PBhjj3K8gN4QjkMvmBD4O46xmbAJ+4X0A3Xnzoy4QFlWR6Vydgk+JWy\n+Sd4xBCnlFKemHiQMV1BoqFaQS50heq6uCosgamGVv+QINqDJjqiBz85Hr2kia529FM79oeK\n0xH8W5F97hK9jW8Xr+DLakjnh7Fd2cyBt5Xl7+EdLV83vJ5orpTd1/L0pdgu6ifwgJbn9XFZ\n8Hnntsp7j+kKEg3VClLRFarvkql++c/APYZW/wjAqWMCRN716wO8SRW9BfLU/XlQvBlm8q26\nKFX0P13jvt+2Irc/jODHf0z55/OwRVmuhJVavm64UbRDFTcHtu+DG+ZzboFd+rgs6AR9angf\nb0GioZ6CFHSF6rtkwgmltQbuYsbX9IcAiK4p6KO8vft+rokuVr/JGHsF1r4Hi9XNn6uiz/DN\nY6NbAFz7CAznoqsYF72D6UXrhhtFt1dXC+HP2zzfUqX6uCyIHQq/4X28BYmGagWp6ArVd3FV\nWANphlb/CMwNyz9XpkPijy7Rf4XJ6v558NFGeE7dbOMRfaUXTNlazU75Fq0bbhTtVL93Z8Cn\ne2CMu1EflwUbvu8QW+HaoRUkGGoQrStU38UrWt/qH/JFH/1FCf9nBvyNreGiqxx96nhDGhw7\n6biLb/0DPKJ3cSX8TuQeX6J1w+udo3fz1X9E//BjZFf1DPHqi2f1cfxr/hY/h+gKEgw1iNYV\nqu/iFa1v9Q/5oo87bvlJqeaWiG/Z+/Ay40qXMa5uEGMj4UPGLqZ7Re+F/1SW5++Eu32J1g83\nih6s3Iy/BY8xlg1zlIbtETfU6eO46Lo74V2mK0gw1CBaX6iui1e0vtU/AnDqmA09ZuTcDDmM\nfQbxQ8tYRScYnDMcOiqXrYfbOu6bdmMi3OE5ddwCwxfndm4V28enaN1wo+j2N84a7eiuXMie\n6gb9c8ZExewwxKlnsf0tk07rChIMNYrWFarrohOta/WPAIi+8npqm9hbl/Mv9cyEuPWMnZic\nHNVlmnryLs9MiB1ZBkM9x3V8XKeY68ceznBU+hKtG24UvW9kTKcp6ru0albXqOTRfzfGuT4u\nZsNEfUGCoQbRukJ1XXSida3+gfvz6MPq/d4J9brfJq6DDtRQiYVq4Ipu1+0i4++yP9iPCqxo\niYVq4Ip+Bq6bmad8VNU23rUxAitaYqEauKLrVvRvG3tz/qXGezZKYEVLLFSDfmeIBIlGgkQj\nQaKRINFIkGgkSDQSdkX/eI4Q8pMk0ZOfnEL4ZkLq72WJPmczILw5RaJxINFIkGgkSDQSJBoJ\nEo0EiUaCRCNBopEg0UiQaCQCJPoRNzZDwwcSjQSJRoJEI0GikSDRSJBoJEg0Emiim7t6Eo1E\noEU3xObLhCokGgkSjQSJRoJEI0GikSDRSJBoJEg0EiQaCRKNBL7oZiqcRCNBopEg0UiQaCRI\nNBJBILp5qCfRSJBoJJpOtEB9OEKikSDRSJBoJEg0Ev6L/iJBWRTw//X3Unb4rtjuf2bulQsS\nLcRf0XUr2zmV1YPPVlZWnr/S6/mLH7U6pK20HiRaiL+iF/Qu4KI7je5521q2M+4yY2NztJXW\ng0QL8Vf08brdiuhv4IWjq2OLl6cqLfn3aCtlsWzRokWPk2gR/p+juWjGJ1zLHbZ0kLIquFNb\nKYt7U1NTs0m0CIuiOW+kvHGrslo8QltpzXTqEGJN9PreykZOxs6EK8rJOU9baftJtBBroqva\nLbmypfXW2huev7Q5br+20vaTaCEWTx2f3R7XfTVjR4bE93jPs3JBooXQnSESJBoJEo0EiUaC\nRCNBopEg0UiQaCRINBIkGgkSjQSJRoJEI0GikSDRSAST6LA2HpSiw1E4iUaCRCNBopEg0UiQ\naCRINBIkGgkSjQSJRoJEI0GikSDRSJBoJIJadDgZJ9FIkGgkSDQSJBoJEo0EiUaCRCNBopEg\n0UiQaCRINBKhIToMjJuL/uEqkki0EHPRcU98bDmJRAsxF724J/RY+P/WklBEh65xX+foz6e0\nbTF07QULSSRaiO8Pw0slaZAw7Su/k0i0EJ+iKxbeDD1zb4pa628SiRZiLvrsm4MccU/sZKxu\nZEd/k0i0EHPR0TDgzfPq1vxr/E0i0ULMReeVK4ufLimLf9X4m0SihZiLrp07iLHN8YvqLCSR\naCHmohe2fI6xyrzIAgtJJFqIuehuL6mrhTdaSCLRQnx8GJaqq1Jng/6+IdFCzEWnzFBXuTdY\nSCLRQsxFFzqe2rTvL1Na0DlaGj7uDOc5ASB6npUkEi3E1y34+a1/3HzWUhKJFuJLdPWxCo6F\nJBItxFz0oQHgwkISiRZiLnpo91ffeZeja1OnBzHOCyJ5epBmKDr6k/r9XNODGOcFkT09SDMU\n3a68fj/X9CDGeUFkTw/SDEVPmFW/n2t6EOO8IN7pQY6UlZVNItEizEW/6LztqZkcXRsXbZwX\nRPb0IM1QdJIbXRsXbZwXxDs9yMaioqLxJFqEtckUjPOCyJ4epFmK3v+b2cf+arg15KKN84LI\nnh6kGYquGafcrOxO66p/hkadHsQ4L4jk6UGaoeh5MW9/D7sPdh1nIQlXdMgJNxd97RJWA7vZ\nKr9/Bc5IdCOYi47coIoO4t+whInolPGq6FkpFpJItBBz0W/AQ+tg1ZyI31lIItFCfFze5cco\nlx2Rc6wkkWghvq6jT5esKv7OUhKJFmIuusqNhSQSLcRcNECw/4YlTEQv5/z23mtKLCSRaCHC\nHyplP24hiUQLEYouSrCQRKKFCEXPJ9HSMBf9LCcvq8VYC0kkWoi56AROm59PPWMhiUQLCbG/\nBQ830VU6/E0i0UIauWGxcM9CooWYi17TesiK0tUPtHhBuW/xN4lECzEXnT5RXU0eaSGJRAsx\nFx2zU11tiLaQRKKFmItOXqqucrtZSGoa0SFj3Fz03OgFe775bLojeP+GJUxE1zzpUK43nAus\nJJFoIb5uWE4UvVlcbSmJRAvx/5GwxiDRQvx/JKwxSLSQUH8kLMRFh84jYSEuOnQeCQtx0aHz\nSFiIiw6dR8JCXHToPBIW6qJD5pGwEBeds9V6EokW4uNPlF+3nkSihZiLHjix1nISiRZiLnpB\nTM+Hc3IVLCSRaCHmotu7sZBEooWE+nMdoSy68O9XlUSihZiIhkLlMjr3oNUkEi3Eh+gKWGc1\nqYlFB7txEo0EiUaCRCNBopEwEx3VqlUsRLfiWEgi0UJMRD+mw0ISiRYSNneGJJpEq5BoJMJP\ntBubhyMbEo0EiUaCRCNBopEg0UiQaCRINBIkGgkSjYRF0QX8z/CXBsH0II0j05IELIp+8NnK\nysrzQTA9SONI1WQfi6I7je5521oWBNODNI5UTfaxJvobeOHo6tjiBtODcEi0EIvv6AvKf7nD\nGkwPMv3RRx8dS6JFXMVVxxspDaYHIdGNYk30+t7KIicjCKYHaRypmuxjTXRVuyVXtrTeGgTT\ngzSOXE+2sXjq+Oz2uO6rg2F6kMaRaUkCdGeIBIlGgkQjQaKRINFIhK9oDzaPSxIkGgkSjQSJ\nRoJEI0GikSDRSJBoJEg0EiQaCRKNBIlGgkQjQaKRaAaiPdg8QHuQaCRINBIkGgkSjQSJRoJE\nI0GikSDRSJBoJJqTaA82j/SqINFIkGgkSDQSJBoJEo0EiUaCRCNBopFolqLd2DxiS5BoJEg0\nEiQaCRKNBIlGgkQjQaKRINFIkGgkSDQSJBqJZi3ag81D9wcSzbF56P5Aojk2D90fSDTH5qH7\nA4nm2Dx0fyDRBmw6EECiDdh0IIBEG7DpQACJNmDTgQASbY5NGQ0h0ebYlNEQEm2OTRkNIdHm\n2JTREFuig3oeFjnYs6vDjujgnodFDnb9erAjOrjnYZFMU4r2zsOysaioaHx4i24InmjvPCz3\npqamZp9rrH+zxo5o7zwsR8rKyiaRaBG2ztG+52Eh6mNHtGAeFqI+tq6jfc/DQtQnQHeGRH1I\nNBIkGgkSjQSJRoJEI0GikZAo+skphG8mSBM9N8PFgAEZgWBovxEBye2XFpDYtH731WvJ/FiS\naDfZ2ZKCjBSlHghI7q0LAhKbn3rZ1y4SLRMSXZ/QFf3115KCjJwuuxCQ3LLjAYk9WVbna5cs\n0UQjkGgkSDQSkkQbHqqRxRcJgUje3j+u8y/rpOcW3xzX/XeCcuWINj5UI4e6le2cAUiuSnjz\n8pddC2Xnfh31Edsb86nvWDmijQ/VyGFB7wJnAJL3Pq4scrKl555m7OPWX/qOlSPa+1CNPI7X\n7XYGJpldTimQn3uuNeQKypUj2vtQjUy46EAkX3xgyCX5ubW1Bzou8x0rR7T3oRqZcNEBSK68\n7eGLgal47kDfsZLO0YaHamTBRctP3tspn8nPLe2rLHIyfMfKEW18qEYWXLT05BMd8isrK6tl\n51YnLrm8NX6971hJ19GGh2pkwUVLT14AnHTpubvviEt5R1Au3RkiQaKRINFIkGgkSDQSQS96\nOL9IiLx2UrXZzuSnTcdkLApoSVdF8ItOLCwsfGVCi9vNfklkLvpS3BcBLuoqCH7RPdTVHPjU\nZKe56E0dff7mrukIFdGbYA1jy1OiOs+vYezHZ3pExQ8/4BK9DF5hfxua0CrN/agKm/2EtnFl\nfufoYR9CuXdk0osT28eMOMK8LViEiugC2MnyYXzJkhjFYlab1za/ntRHFf2W42V2rkP6hpK7\nY09qQ278QNuYHjm/5LFoRbRnZJJzRtmOrncwbwsWwS+6S0VFxb7X2vS6fDrmIeXfb8P+i2kr\nlI2X4bQi+v2IlxjbBR8xdiLnsGvE1y3PuDaOR/xKWWZAuWckS+qnbCyFam8LFsEvWv3ZBPTc\nx0pgXU1NzffwstL63ScrhsApltw3Unlfsx86Jkz+03n3iP8ZqG2s4ScN9gGUe0cmPaW0vAPH\n9Fk4BL/ojuvWFZfyB1NWuZTDbLYxBZypQ+AkS4ZR8LbS6avxHcA5TlM9Ml8bqrxzleUOKPeO\nTJqptLwLFd4WLIJfdA/3VjGs2s05dsg5cv8V9r9cdBZLb/8t31m3J8fxC7Xbxdi92oAP4Etl\nWQTlnpEe0d4WLEJI9LeRc5TlwbQta5UPRsYy4Tj/MPwm9mFWksgfSOsyRu1Weo17QGUkP0dn\nQ7lnpEe0twWLEBLNnouYXbqqa5ez5RHpn2wb54DD6uXdS7C+ukPf9zY9DevUXtPHewZMj1yw\ncWoEHPKM9Ij2tmARSqLrlvWKShqjfLe/3ys6OWsrrFBF1/S+9l/7RiY6+65x9bruQ8+Ay890\ndA7PV975npEe0Z4WLIJetA3OvcWvrOfH1zZ1IZxwFl3zswFFmxdHz23qOlTCWTTbP6q9M+XV\n4PjBR1iLDiZINBIkGgkSjQSJRoJEI0GikSDRSJBoJP4Nb5xvLfAkWZMAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width=3, repr.plot.height=3)\n",
    "# plot a histogram showing peaks per gene\n",
    "p1 <- qplot(table(subjectHits(ov)), binwidth = 1) + theme(plot.subtitle = element_text(vjust = 1), \n",
    "                                                          plot.caption = element_text(vjust = 1)) +\n",
    "  labs(title = \"Histogram of peaks per gene\",  x = \"Peaks / gene\", y=\"Frequency\") + pretty_plot()\n",
    "p1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate distance decay for the weights\n",
    "dist <- abs(mcols(tss_idx)$tss[subjectHits(ov)] - start(atacgranges)[queryHits(ov)])\n",
    "exp_dist_model <- exp(-1*dist/5000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare an outcome matrix\n",
    "m <- Matrix::sparseMatrix(i = c(queryHits(ov), length(atacgranges)),\n",
    "                          j = c(subjectHits(ov), length(tss_idx)),\n",
    "                          x = c(exp_dist_model,0))\n",
    "colnames(m) <- gdf$V4 # gene name\n",
    "m <- m[,which(Matrix::colSums(m) != 0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "fm_genescoring <- data.matrix(t(m) %*% counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>20937</li>\n",
       "\t<li>5335</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 20937\n",
       "\\item 5335\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 20937\n",
       "2. 5335\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 20937  5335"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A matrix: 3 × 3 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>AAACGAAAGCGCAATG-1</th><th scope=col>AAACGAAAGGGTATCG-1</th><th scope=col>AAACGAAAGTAACATG-1</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>A1BG</th><td>0.0862008631</td><td>1.7756937</td><td>3.591624</td></tr>\n",
       "\t<tr><th scope=row>A1BG-AS1</th><td>0.4582008458</td><td>0.4783511</td><td>1.365454</td></tr>\n",
       "\t<tr><th scope=row>A2LD1</th><td>0.0003629473</td><td>0.0000000</td><td>0.000000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 3 × 3 of type dbl\n",
       "\\begin{tabular}{r|lll}\n",
       "  & AAACGAAAGCGCAATG-1 & AAACGAAAGGGTATCG-1 & AAACGAAAGTAACATG-1\\\\\n",
       "\\hline\n",
       "\tA1BG & 0.0862008631 & 1.7756937 & 3.591624\\\\\n",
       "\tA1BG-AS1 & 0.4582008458 & 0.4783511 & 1.365454\\\\\n",
       "\tA2LD1 & 0.0003629473 & 0.0000000 & 0.000000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 3 × 3 of type dbl\n",
       "\n",
       "| <!--/--> | AAACGAAAGCGCAATG-1 | AAACGAAAGGGTATCG-1 | AAACGAAAGTAACATG-1 |\n",
       "|---|---|---|---|\n",
       "| A1BG | 0.0862008631 | 1.7756937 | 3.591624 |\n",
       "| A1BG-AS1 | 0.4582008458 | 0.4783511 | 1.365454 |\n",
       "| A2LD1 | 0.0003629473 | 0.0000000 | 0.000000 |\n",
       "\n"
      ],
      "text/plain": [
       "         AAACGAAAGCGCAATG-1 AAACGAAAGGGTATCG-1 AAACGAAAGTAACATG-1\n",
       "A1BG     0.0862008631       1.7756937          3.591624          \n",
       "A1BG-AS1 0.4582008458       0.4783511          1.365454          \n",
       "A2LD1    0.0003629473       0.0000000          0.000000          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(fm_genescoring)\n",
    "fm_genescoring[1:3,1:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "end_time <- Sys.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Time difference of 12.63262 secs"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "end_time - start_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all(colnames(fm_genescoring) == rownames(metadata))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(fm_genescoring, file = '../../output/feature_matrices/FM_GeneScoring_10xpbmc5k.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 3.5.1 (2018-07-02)\n",
       "Platform: x86_64-conda_cos6-linux-gnu (64-bit)\n",
       "Running under: CentOS Linux 7 (Core)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /data/pinello/SHARED_SOFTWARE/anaconda3/envs/ATACseq_GeneScoring/lib/R/lib/libRblas.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] parallel  stats4    stats     graphics  grDevices utils     datasets \n",
       "[8] methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] Matrix_1.2-17               BuenColors_0.5.5           \n",
       " [3] ggplot2_3.2.0               MASS_7.3-51.4              \n",
       " [5] dplyr_0.8.3                 data.table_1.12.2          \n",
       " [7] SummarizedExperiment_1.12.0 DelayedArray_0.8.0         \n",
       " [9] BiocParallel_1.16.6         matrixStats_0.54.0         \n",
       "[11] Biobase_2.42.0              GenomicRanges_1.34.0       \n",
       "[13] GenomeInfoDb_1.18.1         IRanges_2.16.0             \n",
       "[15] S4Vectors_0.20.1            BiocGenerics_0.28.0        \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] Rcpp_1.0.1             pillar_1.4.2           compiler_3.5.1        \n",
       " [4] XVector_0.22.0         bitops_1.0-6           base64enc_0.1-3       \n",
       " [7] tools_3.5.1            zlibbioc_1.28.0        zeallot_0.1.0         \n",
       "[10] digest_0.6.20          uuid_0.1-2             gtable_0.3.0          \n",
       "[13] tibble_2.1.3           jsonlite_1.6           evaluate_0.14         \n",
       "[16] lattice_0.20-38        pkgconfig_2.0.2        rlang_0.4.0           \n",
       "[19] IRdisplay_0.7.0        IRkernel_1.0.1         GenomeInfoDbData_1.2.1\n",
       "[22] withr_2.1.2            repr_1.0.1             vctrs_0.2.0           \n",
       "[25] tidyselect_0.2.5       grid_3.5.1             glue_1.3.1            \n",
       "[28] R6_2.4.0               pbdZMQ_0.3-3           purrr_0.3.2           \n",
       "[31] magrittr_1.5           backports_1.1.4        scales_1.0.0          \n",
       "[34] htmltools_0.3.6        assertthat_0.2.1       colorspace_1.4-1      \n",
       "[37] labeling_0.3           lazyeval_0.2.2         munsell_0.5.0         \n",
       "[40] RCurl_1.95-4.12        crayon_1.3.4          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "save.image(file = 'GeneScoring_10xpbmc5k.RData')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:ATACseq_GeneScoring]",
   "language": "R",
   "name": "conda-env-ATACseq_GeneScoring-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
